src/meta/src/hummock/compactor_manager.rs

```diff
@@ -82,164 +92,199 @@ impl Compactor {
     }
 }
 
-struct CompactorManagerInner {
-    /// Senders of stream to available compactors
-    compactors: Vec<HummockContextId>,
-
-    /// TODO: Let each compactor have its own Mutex, we should not need to lock whole thing.
-    /// The outer lock is a RwLock, so we should still be able to modify each compactor
-    compactor_map: HashMap<HummockContextId, Arc<Compactor>>,
-
-    /// We use round-robin approach to assign tasks to compactors.
-    /// This field indexes the compactor which the next task should be assigned to.
-    next_compactor: usize,
-}
-
-impl CompactorManagerInner {
-    pub fn new() -> Self {
-        Self {
-            compactors: vec![],
-            compactor_map: HashMap::new(),
-            next_compactor: 0,
-        }
-    }
-}
-
 /// `CompactorManager` maintains compactors which can process compact task.
 /// A compact task is tracked in `HummockManager::Compaction` via both `CompactStatus` and
-/// `CompactTaskAssignment`. A compact task can be in one of these states:
-/// - 1. Assigned: a compact task is assigned to a compactor via `HummockManager::get_compact_task`.
-///   Assigned-->Finished/Cancelled.
-/// - 2. Finished: an assigned task is reported as finished via
-///   `CompactStatus::report_compact_task`. It's the final state.
-/// - 3. Cancelled: an assigned task is reported as cancelled via
-///   `CompactStatus::report_compact_task`. It's the final state.
+/// `CompactTaskAssignment`.
+///
+/// A compact task can be in one of these states:
+/// - 1. Pending: a compact task is picked but not assigned to a compactor.
+///   Pending-->Success/Failed/Canceled.
+/// - 2. Success: an assigned task is reported as success via `CompactStatus::report_compact_task`.
+///   It's the final state.
+/// - 3. Failed: an Failed task is reported as success via `CompactStatus::report_compact_task`.
+///   It's the final state.
+/// - 4. Cancelled: a task is reported as cancelled via `CompactStatus::report_compact_task`. It's
+///   the final state.
+/// We omit Assigned state because there's nothing to be done about this state currently.
+///
+/// Furthermore, the compactor for a compaction task must be picked with `CompactorManager`,
+/// or its internal states might not be correctly maintained.
 pub struct CompactorManager {
-    inner: parking_lot::RwLock<CompactorManagerInner>,
+    // `policy` must be locked before `compactor_assigned_task_num`.
+    policy: RwLock<Box<dyn CompactionSchedulePolicy>>,
+    // Map compactor to the number of assigned tasks.
+    compactor_assigned_task_num: Mutex<HashMap<HummockContextId, u64>>,
```

Comment:
Actually, should the number of assigned task stored in the `Compactor` struct since it is part of a compactor's state?

Reply:
> Actually, should the number of assigned task stored in the `Compactor` struct since it is part of a compactor's state?

Sound good to me, we may collect more information of compactor. Hence, we can stored the `State` in `Compactor`

Reply:
LGTM as well.

Reply:
One small problem: when `Policy` is recovering the state from assigned tasks, `Compactor` does not exist at that time because the corresponding compactor might not have subscribed to meta, so we must use another map to store the number of assigned tasks, and use that to initialize the `Compactor ` when it subscribes. And we have to repeat this logic in every implementation of `CompactionSchedulePolicy`.

Reply:
I purpose that we redefine `Compactor`

```rust
struct Compactor {
    context_id: HummockContextId,
    /// The inherent state of a compactor (compators that have not subscribed can also have this).
    state: CompactorState,
    /// The information of the compactor that becomes available only after subscription.
    subscription_info: Option<CompactorSubscriptionInfo>,
}

struct CompactorState {
    assigned_task_num: usize,
}

struct CompactorSubscriptionInfo {
    sender: Sender<MetaResult<SubscribeCompactTasksResponse>>,
    max_concurrent_task_number: u64,
}
```

When choosing compactors, compactors without `CompactorSubscriptionInfo` will be skipped.

Reply:
With compactor picking moved into `assign_compaction_task`, there's no need to maintain this state in `CompactorManager`.